lib/active_model/serializer/type.rb

```diff
@@ -16,8 +16,13 @@ module ClassMethods
         # @example
         #   class AdminAuthorSerializer < ActiveModel::Serializer
         #     type 'authors'
+        #
+        # @example
+        # TODO: actually do block, not Proc
+        #   class AdminAuthorSerializer < ActiveModel::Serializer
+        #     type { |object| object.class.name }
         def type(type)
-          self._type = type && type.to_s
+          self._type = type.is_a?(Proc) ? type : type.to_s if type
```

Comment:
see [CollectionSerializer#json_key](https://github.com/rails-api/active_model_serializers/pull/1618/files#diff-78e515848bf3a42cc6aa1e892f6798caR30) for some more discussion about what I think would be better to do first